server/controllers/login.js

```diff
@@ -0,0 +1,131 @@
+        if (!scope.includes('read:org') || !scope.includes('user:email')){
+            return res.status(401).json({
+                message: 'More permissions are required'
+            })
+        }
+
+        //Fetch user email from github
+        const ghUserEmailResponse = await axios({
+            method: 'get',
+            url: `${constants.githubAPIDomain}/user/emails`,
+            headers: {
+                accept: 'application/vnd.github.v3+json',
+                Authorization: `token ${ghTokenResponse.data.access_token}`
+            }  
+        })
+
+        //Check if the user exists in our db
+        let user = await User.findOne({email: ghUserEmailResponse.data[0].email})
+        if(!user){
+            //Create a new user if the user does not exist in our db
+            const newUser = {
+                'email': ghUserEmailResponse.data[0].email
+            }
+            //Fetch user info from github
+            const ghUserResponse = await axios({
+                method: 'get',
+                url: `${constants.githubAPIDomain}/user`,
+                headers: {
+                    accept: 'application/vnd.github.v3+json',
+                    Authorization: `token ${ghTokenResponse.data.access_token}`
+                }  
+            })
+            newUser['name'] = ghUserResponse.data.name || ghUserResponse.data.login
+            newUser['username'] = ghUserResponse.data.login
+            newUser['avatarUrl'] = ghUserResponse.data.avatar_url
+
+            /*
+                In order to handle a case when user tries signing in gets created on RC but fails get stored in our db due to some error
+                To keep it consistent with RC we store the (user-email,encrypted-rcPassword) pair in our fs so that it can be used
+                to store later in our db
+            */
+            let inconsistentUsers = jsonfile.readFileSync(inconsistentUsersPath)
+            let rcPassword = inconsistentUsers[newUser.email]
+            
+            if(rcPassword){
+                //If the user already exists on RC
+                delete inconsistentUsers[newUser.email]
+                jsonfile.writeFileSync(inconsistentUsersPath, inconsistentUsers)
+            } else {
+                //Generate RC password and create user on RC
+                rcPassword = Math.random().toString(36).slice(2) +  Math.random().toString(36).toUpperCase().slice(2)
+                await axios({
+                    method: 'post',
+                    url: `${constants.rocketChatAPIURL}/users.register`,
+                    headers: {
+                        'Content-type': 'application/json'
+                    },
+                    data: `{
+                        "name": "${newUser.name}",
+                        "email": "${newUser.username}@rc4git.com",
+                        "pass": "${rcPassword}",
+                        "username": "${newUser.username}_rc4git"
```

Comment:
```suggestion
                        "username": "${newUser.username}_rc4git_github"
```

Reply:
@murtaza98 The above-suggested change is great and in order to make an account on RC, I'll make the change as you have suggested.

This made me give a deeper thought on the scenarios/edge case which could arise. Let say if there are two users, Tom and Jerry (needless to say both will be having different email ids), and Tom has a Github username which Jerry has as a Gitlab username. In such a case rc4git usernames might clash and since the username field is set to be unique it would throw an error.

What do you think of setting the rc4git username as `${username}_${remote_vcs}`? E.g. if a user with a username as tom logs-in with github then his username would be tom_github on rc4git. 

We can extend this for the above case as well i.e. for RC, we can make it as username_github_rc4git.

Please let me know your thoughts.
(Also we'll be surely giving the user an option to select his/her username but there we can easily handle the case of username duplication if someone tries to take the username which has been already taken up on rc4git)

Reply:
@djcruz93 Nice catch! Since we are supporting multiple platforms, so we need to handle this kind of edge cases. I'm onboard with this format for rc4git username
`${username}_${remote_vcs}`

And wrt the rocket.chat username I think you have already got it very well, i.e 
`${username}_${remote_vcs}_rc4git`

Reply:
Great!